geoproxy/main.go

```diff
@@ -95,42 +113,117 @@ func parseRequestBody(request *http.Request) requestPayloadStruct {
+		// Returning error coord of 360, 360 for now
+		log.Printf("Error: Coordinate passed into json request should be of form [0.0, 0.0].")
+		return Coord{360, 360}
+	} else {
+		coord0 := []rune(splitCoord[0])
+		coord1 := []rune(splitCoord[1])
+
+		lat, errLat := strconv.ParseFloat(string(coord0[1:len(splitCoord[0])]), 64)
+		lng, errLng := strconv.ParseFloat(string(coord1[0:len(splitCoord[1])-1]), 64)
+
+		if errLat != nil || errLng != nil {
+			log.Printf("Error: Invalid lat-long coordinate.")
+			return Coord{360, 360}
+		}
+		return Coord{lat, lng}
 	}
+}
+
+func rangesOverlap(start1 float64, end1 float64, start2 float64, end2 float64) bool {
+	return math.Min(end1, end2) >= math.Max(start1, start2)
+}
 
-	// Lat-long mapping logic to get the env
-	for l1 := range sorted_lat {
-		if lat < float32(l1) {
-			var sorted_lng []int
-			for l2 := range data[l1] {
-			    sorted_lng = append(sorted_lng, l2)
+// Get the url(s) for given coordinates of view request
+func getViewProxyUrl(rawCoord1 string, rawCoord2 string) map[string]bool {
+	// Acts as a set of urls
+	urls := make(map[string]bool)
+
+	// Parse each coord
+	topLeft := parseCoord(rawCoord1)
+	bottomRight := parseCoord(rawCoord2)
+
+	// **Commented out to test with only one server associated with midpoint for now**
+	// *******************************************************************************
+	// // Add all urls of zones that touch the box of the coords
+	// for latRange := range data {
+	// 	if rangesOverlap(topLeft.Lat, bottomRight.Lat, latRange.Low, latRange.High) {
+	// 		for lngRange := range data[latRange] {
+	// 			if rangesOverlap(topLeft.Lng, bottomRight.Lng, lngRange.Low, lngRange.High) {
+	// 				// Check if we have already added url
+	// 				urlString := os.Getenv(data[latRange][lngRange])
+	// 				// url, exists := urls[urlString]
+	// 				if !urls[urlString] {
+	// 					urls[urlString] = true
+	// 				}
+	// 			}
+	// 		}
+	// 	}
+	// }
+	// *******************************************************************************
+
+	coord := Coord{(topLeft.Lat + bottomRight.Lat) / 2, (topLeft.Lng + bottomRight.Lng) / 2}
+
+	for latRange := range data {
+		if coord.Lat >= latRange.Low && coord.Lat < latRange.High {
+			for lngRange := range data[latRange] {
+				if coord.Lng >= lngRange.Low && coord.Lng < lngRange.High {
+					urls[os.Getenv(data[latRange][lngRange])] = true
+					break
+				}
 			}
-			sort.Ints(sorted_lng)
-			for l2 := range sorted_lng {
-				if lng < float32(l2) {
-					return os.Getenv(data[l1][l2])
+			break
+		}
+	}
+
+	return urls
+}
+
+// Get the url for given coordinates of submit request
+func getSubmitProxyUrl(rawCoord string) string {
+	coord := parseCoord(rawCoord)
+	// Lookup coord in data map
+	for latRange := range data {
+		if coord.Lat >= latRange.Low && coord.Lat < latRange.High {
+			for lngRange := range data[latRange] {
+				if coord.Lng >= lngRange.Low && coord.Lng < lngRange.High {
+					return os.Getenv(data[latRange][lngRange])
```

Comment:
Seems to duplicate code on line 198 (for the most part). Can we extract into a function?